src/ray/raylet/local_object_manager.cc

```diff
@@ -265,6 +295,86 @@ void LocalObjectManager::AsyncRestoreSpilledObject(
   });
 }
 
+void LocalObjectManager::ProcessSpilledObjectsDeleteQueue(uint32_t max_batch_size) {
+  RAY_CHECK(object_pinning_enabled_) << "Not Implemented";
+  absl::MutexLock lock(&mutex_);
+  std::vector<std::string> object_urls_to_delete;
+
+  // Process upto batch size of objects to delete.
+  while (!spilled_object_pending_delete_.empty() &&
+         object_urls_to_delete.size() < max_batch_size) {
+    auto &object_id = spilled_object_pending_delete_.front();
+    // If the object is still spilling, do nothing. This will block other entries to be
+    // processed, but it should be fine because the spilling will be eventually done, and
+    // deleting objects is the low priority tasks.
+    // This will instead enable simpler logic after this block.
+    if (objects_pending_spill_.contains(object_id)) {
+      break;
+    }
+
+    // Object id is either spilled or not spilled at this point.
+    const auto spilled_objects_url_it = spilled_objects_url_.find(object_id);
+    if (spilled_objects_url_it != spilled_objects_url_.end()) {
+      // If the object was spilled, see if we can delete it. We should first check the ref
+      // count.
+      std::string &object_url = spilled_objects_url_it->second;
+      // Note that here, we need to parse the object url to obtain the base_url.
+      auto parsed_url = ParseURL(object_url);
+      const auto base_url_it = parsed_url->find("url");
+      RAY_CHECK(base_url_it != parsed_url->end());
+      const auto &url_ref_count_it = url_ref_count_.find(base_url_it->second);
+      RAY_CHECK(url_ref_count_it != url_ref_count_.end())
+          << "url_ref_count_ should exist when spilled_objects_url_ exists. Please "
+             "submit a Github issue if you see this error.";
+      url_ref_count_it->second -= 1;
+
+      // If there's no more refs, delete the object.
+      if (url_ref_count_it->second == 0) {
+        url_ref_count_.erase(url_ref_count_it);
+        object_urls_to_delete.emplace_back(object_url);
+      }
+      spilled_objects_url_.erase(spilled_objects_url_it);
```

Comment:
Hmm how does this work, doesn't this data structure need to be available on all raylets?

Reply:
Answered in other thread.